### ISSUE TITLE
invalid/non stable ast for combinator with comment

### ISSUE THREAD

**Issue by evilebottnawi** [postcss/postcss-selector-parser#189]
Input:
```
h2 /*test*/ h4 { } // Here we don't have `comment` node in ast (store comment in `raws`)
h2/*test*/h4 { } // Comment in ast
```

It is do impossible analyzes comments, also adopt new version to `stylelint` and `prettier`. For me it is bug, but fixing this change ast.

**Comment by ai** [comment]
Why it will change AST?

**Comment by evilebottnawi** [comment]
@ai new `comment` node appears in ast(before it was in `raws`), just want to clarify should we release this as patch or major?

**Comment by ai** [comment]
Technically it should be major

**Comment by evilebottnawi** [comment]
Same problem for `[ /*t*/ title /*t*/ = /*t*/ "Something" /*t*/ ]`, comments should be part of ast not `raw`

**Comment by evilebottnawi** [comment]
/cc @ai i ahve some questions about spaces too
Example:
```
[   href   =   "test"   i   ] { }
  ^       ^    ^          ^  ^
```

What node should store spaces in this case (inside `raws`?)? Because now logic in parser is very misleading and I can't understand whether we do it right or not

**Comment by ai** [comment]
To be honest, I am not sure that I know the best answer.

But I think in selector parser we should work with spaces in a different way, compare to PostCSS core. In main CSS syntax, whitespaces don’t mean anything. In selector, whitespace can have a lot of meanings.

So, I think it could be better to have a special token for spaces.

**Comment by evilebottnawi** [comment]
@ai yep, we already have special token for `space`, my questions about where we should store meta information about space (`raws`). Let's see on example above:
- First spaces should be stored in `raws.before` or `raws.attribute.before`?
- Second space should be stored in `raws.attribute.before` or `raws.operatore.before`?
- Third space should be stored in `raws.operator.after` or `raws.value.before`
- Four space should be stored in `raws.value.after` or `raws.insensitive.before` (`insensitive` will be renamed in `flag` in next major)
- Five space should be stored in `raws.insensitive.after` or `raws.after` (`insensitive` will be renamed in `flag` in next major)

**Comment by ai** [comment]
Why do we need a `raws` for whitespaces when we have `space` token?

**Comment by evilebottnawi** [comment]
@ai hm can you provide example/PoC of ast with `space` token? We need this spaces for linting/printing in `stylelint` and `prettier`

**Comment by ai** [comment]
`a b` > `{ type: "word", value: "a" }, { type: "space", value: " " }, { type: "word", value: "b" }`

**Comment by evilebottnawi** [comment]
@ai in example above `space` is `descendant selector`, we already provide `combinator` as separate node, but in my example spaces mean nothing so we omit them from ast

**Comment by ai** [comment]
I am not sure that we should omit them from AST.

**Comment by evilebottnawi** [comment]
@ai I do not see their meaning as they really do not carry here any semantic load, also we do this right now, we can plan this on next major, but will be great solve problems with spaces using `raws` for next release

**Comment by ai** [comment]
Yeap. This discussion is on you. Sorry, that I can't help right now.

**Comment by evilebottnawi** [comment]
@ai Maybe not related to issue, but what is blocker integrate selector parser in postcss? Non standard syntax? Or nobody send a PR?

**Comment by ai** [comment]
@evilebottnawi I am afraid that we will freeze API which will not be the best.

**Comment by evilebottnawi** [comment]
@ai we can afraid this forever :smile: To be honestly ast of `postcss` is not enough for prettier/stylelint nowadays  (prettier breaks code in many cases and i recommended don't use this for css/scss/less/etc). I think we should start integration `selector`/`value`/`media` parsers otherwise we will need do fork postcss and continue development :disappointed:

**Comment by evilebottnawi** [comment]
Also we can release this under flag/option and as experimental (adding information to readme about non stable)

**Comment by ai** [comment]
RIght now our main focus is visitor API https://github.com/postcss/postcss/pull/1245

**Comment by evilebottnawi** [comment]
@ai Is there any sense send a PR or better fork postcss and starting own development for prettier?

**Comment by ai** [comment]
@evilebottnawi I have a better plan. Let’s release new AST in `postcss-selector-parser` and if nobody will complain about it, we can move it to PostCSS

**Comment by evilebottnawi** [comment]
@ai okay :+1: i will ping you before release or when i have some questions
